Sources/Aperture/Aperture.swift

```diff
@@ -1,228 +1,823 @@
+				return buffer
+			}
+			if lastFrame.flags.contains(.valid) {
+				if timeOffset.value > 0 {
+					pts = CMTimeSubtract(pts, timeOffset)
+				}
+
+				let offset = CMTimeSubtract(pts, lastFrame)
+				if timeOffset.value == 0 {
+					timeOffset = offset
+				} else {
+					timeOffset = CMTimeAdd(timeOffset, offset)
+				}
+			}
+			self.lastFrame?.flags = []
+		}
+
+		if timeOffset.value > 0 {
+			resultBuffer = resultBuffer.adjustTime(by: timeOffset) ?? resultBuffer
+		}
+
+		var lastFrame = CMSampleBufferGetPresentationTimeStamp(resultBuffer)
+		let dur = CMSampleBufferGetDuration(resultBuffer)
+		if dur.value > 0 {
+			lastFrame = CMTimeAdd(lastFrame, dur)
+		}
+
+		self.lastFrame = lastFrame
+
+		return resultBuffer
+	}
+}
+
+extension Aperture.Recorder: SCStreamDelegate {
+	public func stream(_ stream: SCStream, didStopWithError error: any Error) {
+		recordError(error: .unknown(error))
+	}
+}
+
+extension Aperture.Recorder: SCStreamOutput {
+	public func stream(_ stream: SCStream, didOutputSampleBuffer sampleBuffer: CMSampleBuffer, of type: SCStreamOutputType) {
+		if isPaused {
+			return
+		}
+		guard sampleBuffer.isValid else {
+			return
+		}
+
+		let sampleBuffer = handleBuffer(buffer: sampleBuffer, isVideo: type == .screen)
+
+		switch type {
+		case .screen:
+			guard
+				let attachmentsArray = CMSampleBufferGetSampleAttachmentsArray(sampleBuffer, createIfNecessary: false) as? [[SCStreamFrameInfo: Any]],
+				let attachments = attachmentsArray.first,
+				let statusRawValue = attachments[SCStreamFrameInfo.status] as? Int,
+				let status = SCFrameStatus(rawValue: statusRawValue),
+				status == .complete
+			else {
+			    return
+			}
+
+			if assetWriter != nil && !isRunning {
+				startVideoStream(sampleBuffer: sampleBuffer)
+			}
+
+			if isRunning, let videoInput, videoInput.isReadyForMoreMediaData {
+				videoInput.append(sampleBuffer)
+			}
+		case .audio:
+			if assetWriter != nil && !isRunning && target == .audioOnly {
+				startAudioStream(sampleBuffer: sampleBuffer)
+			}
+
+			if isRunning, let systemAudioInput, systemAudioInput.isReadyForMoreMediaData {
+				systemAudioInput.append(sampleBuffer)
+			}
+		case .microphone:
+			if assetWriter != nil && !isRunning && target == .audioOnly {
+				startAudioStream(sampleBuffer: sampleBuffer)
+			}
+
+			if isRunning, let microphoneInput, microphoneInput.isReadyForMoreMediaData {
+				microphoneInput.append(sampleBuffer)
+			}
+		default:
+			break
+		}
 	}
+}
+
+extension Aperture.Recorder: AVCaptureAudioDataOutputSampleBufferDelegate, AVCaptureVideoDataOutputSampleBufferDelegate {
+	public func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
+		if isPaused {
+			return
+		}
+
+		let sampleBuffer = handleBuffer(buffer: sampleBuffer, isVideo: output is AVCaptureVideoDataOutput)
+
+		if output is AVCaptureAudioDataOutput {
+			if assetWriter != nil && !isRunning && target == .audioOnly {
```

Comment:
```suggestion
			if
				assetWriter != nil,
				!isRunning,
				target == .audioOnly
			{
```

Same in many places. Prefer `,` over `&&` in if/guard statements.